core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -818,6 +859,51 @@ private static AnnotatedValueResolver ofInjectableTypes0(String name, AnnotatedE
         };
     }
 
+    /**
+     * Returns an attribute resolver which retrieves a value specified by {@code attrKeys}
+     * from the {@link RequestContext}.
+     */
+    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
+    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
+        return (resolver, ctx) -> {
+            StringBuilder errorMsgBuilder = null;
+            Class<?> targetType = resolver.rawContainerType();
+            if (targetType == null) {
+                targetType = resolver.elementType().isPrimitive() ?
+                              Primitives.wrap(resolver.elementType())
+                              : resolver.elementType();
+            }
+
+            for (AttributeKey<?> attrKey : attrKeys) {
+                final Object value = ctx.context.attr(attrKey);
+                if (value != null) {
+                    final boolean isValidType = targetType.isInstance(value);
+
+                    if (isValidType) {
+                        return value;
+                    } else {
+                        if (errorMsgBuilder == null) {
+                            errorMsgBuilder = new StringBuilder(
+                                    String.format("No matching value for '%s' in the attributes: [",
+                                                  resolver.httpElementName()));
+                            errorMsgBuilder.append(String.format("{key=%s, value=%s}", attrKey.name(), value));
```

Comment:
Rather than the value itself, I imagine people will be more interested in the class.
Additionally, I think the string manipulation is getting too complicated - most users will probably only be interested in the last non-null value.
What do you think of just notifying the last non-null invalid value instead like suggested in a previous comment?
https://github.com/line/armeria/pull/5547#discussion_r1568345912

e.g.
```java
Object invalidValue = null;
for (AttributeKey<?> attrKey : attrKeys) {
    final Object value = ctx.context.attr(attrKey);
    if (value != null) {
        final boolean isValidType = targetType.isInstance(value);

        if (isValidType) {
            return value;
        } else {
            invalidValue = value;
        }
    }
}

if (invalidValue != null) {
    throw new IllegalStateException(String.format("{(%s) is not an instance of (%s)",
                                                  targetType.getSimpleName(),
                                                  invalidValue.getClass().getSimpleName()));
} else {
    return resolver.defaultOrException();
}
```

Reply:
> Additionally, I think the string manipulation is getting too complicated 
I agree. 

> What do you think of just notifying the last non-null invalid value instead like suggested in a previous comment?

i think it make sense as well. IMO, it would be better to add `AttributeKey` name.
User may want to know where that `value` came from.

Reply:
<img width="963" alt="image" src="https://github.com/line/armeria/assets/90125071/56786e14-b716-4178-a0b6-88de18769a0f">
Quite long string -_-... 
However, it will be helpful to person, i think.